### ISSUE TITLE
Update the `wp_add_inline_script` function to ensure the inline scripts specify a `type="text/template"` attribute for inline scripts that are associated with deferred, when applicable

### ISSUE THREAD

**Issue by 10upsimon** [10up/wordpress-develop#18]
## Description
In order for applicable inline scripts to execute only once their deferred/async main/parent scripts have loaded ([see issue 17](https://github.com/orgs/10up/projects/14/views/1?pane=issue&itemId=20087860)), scripts should not execute automatically, and only on command. For this reason, `type="text/template"` should be added to said inline `<script>` tags. This prevents them from being executed by the browser.

`onload` handlers will be applied to the parent/main script in a future issue, ensuring that the inline module scripts be converted to JavaScript scripts, and then executed.

## Acceptance Criteria

* Enhance the existing [WP_Scripts](https://github.com/WordPress/WordPress/blob/6.1-branch/wp-includes/class-wp-scripts.php) class, converting the printed inline script tags (where applicable) to be of `type="text/template"` as opposed to `type="text/javascript"` when they meet the business criteria as outlined in [issue 17](https://github.com/orgs/10up/projects/14/views/1?pane=issue&itemId=20087860).

### Unit Tests
* Unit tests for the various processes surrounding `wp_add_inline_script` method currently reside in WordPress core source, starting [at this location](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/dependencies/scripts.php#L395).
* Add/edit the necessary unit tests to ensure that inline scripts that meet the criteria as per [issue 17](https://github.com/orgs/10up/projects/14/views/1?pane=issue&itemId=20087860) or of `type="module"`.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by joemcgill** [comment]
I'm not sure if this has already been considered, but another option we could consider here is to load the inline scripts as a callback function on custom browser events that are fired as part of the `onLoad` handler. This way the inline script would be executed after the deferred/async script without the need for DOM manipulation.

**Comment by 10upsimon** [comment]
@joemcgill I'm not sure I understand the suggestion here. Can you provide a reference to or an example of such a custom browser event? If you're referring to a  custom JavaScript event still, how can we avoid any form of DOM manipulation, even if we are to introduce a new script element as opposed to modifying an existing text/template one?

**Comment by joemcgill** [comment]
This may be a totally incoherent suggestion, but instead of printing the inline script into a script with `type="text/template"` that needs to be modified during `onload` handlers, I was thinking we could create [a synthetic event](https://developer.mozilla.org/en-US/docs/Web/Events/Creating_and_triggering_events) and dispatch that event with the handle of the script during via the `onLoad` attribute of the main script. Here's a pseudocode example:

```js
const wpScriptLoaded = new CustomEvent("wp-script-loaded", { detail: {handle: elem.dataset.handle });

// Main script.
<script src="{path}" defer data-handle="main-script" onLoad="() => document.dispatch(wpScriptLoaded)"

// Inline script, wrapped in a listener callback.
<script>
document.addEventListner(
  "wp-script-loaded",
  (e) => {
    if (e.detail.handle === 'main-script' {
      // Inline script printed here:
      ...do stuff.
    }
  }
);
</script>
```

Hopefully that explanation is more clear.

**Comment by 10upsimon** [comment]
Thanks @joemcgill  and this is actually exactly what I was looking at last night, so we are indeed on the same page. I have no objection to this approach, do you @felixarntz ?

Browser support is probably good enough too: https://caniuse.com/?search=CustomEvent

**Comment by felixarntz** [comment]
@joemcgill @10upsimon We had discussed wrapping the inline script in logic at some point a few weeks/months back, but that is not feasible _because_ it means the code is wrapped. We need to ensure the scope of the inline JS execution remains the same (e.g. global JS variables remain global).

That's why we should go with the `text/template` approach - it keeps the script content as is.

**Comment by joemcgill** [comment]
That reasoning makes sense to me. Thanks for confirming @felixarntz.

**Comment by 10upsimon** [comment]
Thanks @felixarntz , leaving this as is then and ready to proceed @kt-12
